### PR TITLE
add config file rendering functionality to kd deploy

### PR DESCRIPTION
**Pull request by AwkwardBen**

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -183,7 +183,18 @@ func run(c *cli.Context) error {
 func render(tmpl string, vars map[string]string) (string, error) {
 	fm := template.FuncMap{
 		"split": strings.Split,
-	}
+                "file": func(key string) string {
+                        data, err := ioutil.ReadFile(key)
+                        if err != nil {
+                                return ""
```
Comment (KashifSaadat) [inline main.go]: You should return empty string and the error with some context, both here and below

File: main.go [hunk 2]
```diff
@@ -183,7 +183,18 @@ func run(c *cli.Context) error {
 func render(tmpl string, vars map[string]string) (string, error) {
 	fm := template.FuncMap{
 		"split": strings.Split,
-	}
+                "file": func(key string) string {
```
Comment (marcinc) [inline main.go]: indentation!

**Comment by AwkwardBen** [comment]
Completed some local testing, setup is test.yaml has the deployment and config.yaml has it's configuration. In the case when file doesn't exist this is the output:

**Comment by gambol99** [comment]
can we just jot down a use-case

File: main.go [hunk 3]
```diff
@@ -186,11 +186,22 @@ func render(tmpl string, vars map[string]string) (string, error) {
 		"hasPrefix": strings.HasPrefix,
 		"hasSuffix": strings.HasSuffix,
 		"split": strings.Split,
-	}
+                "file": func(key string) (string, error) {
+                        data, err := ioutil.ReadFile(key)
+                        if err != nil {
+                                return "" , err
```
Comment (gambol99) [inline main.go]: Its rare to throw and error in a template function ... given the template content is important i'd say it either works or throw's a panic .. In the case return string and above the Execute of the template defer and capture panic, os.Exit(1) with a contextual error .. https://github.com/kubernetes/kops/blob/master/pkg/util/templater/templater.go#L62-L70

Comment (gambol99) [reply main.go]: Also remove the inline function and create a new function out of it

File: main.go [hunk 4]
```diff
@@ -194,7 +195,20 @@ func render(tmpl string, vars map[string]string) (string, error) {
...
 		return b.String(), err
 	}
-	return b.String(), nil
+	// need to replace blank lines because of bad template formating
+	return strings.Replace(b.String(), "\n\n", "\n", -1), nil
+}
+
+func fileRender(key string) string {
+	data, err := ioutil.ReadFile(key)
+	if err != nil {
```
Comment (gambol99) [inline main.go]: move the defer into the render function to cover render() and it's LGTM

File: main.go [hunk 5]
```diff
@@ -194,7 +195,20 @@ func render(tmpl string, vars map[string]string) (string, error) {
 	if err := t.Execute(&b, vars); err != nil {
```
Comment (gambol99) [inline main.go]: so above the Execute

```Go
defer func() {
	if r := recover(); r != nil {
		err = fmt.Errorf("failed to parse template, error: %s", r)
	}
}()
```

**Pull request merged**